### ISSUE TITLE
Inconsistent union type narrowing based on literal attribute

### ISSUE THREAD

**Issue by shiftinv** [microsoft/pyright#3985]
**Describe the bug**
Narrowing a union based on an attribute's literal value does not work consistently.

**To Reproduce**
Create two types (`A`, `B`) with unique literal attributes (`type`), try to narrow `self.x: A | B` in another type using `self.x.type`.

**Expected behavior**
The attribute is correctly narrowed to the type dictated by the literal value.

**Screenshots or Code**
```python
from typing import Literal

class A:
    type: Literal[0]

class B:
    type: Literal[1]

class X:
    thing: A | B

    def test(self) -> None:
        if self.thing.type == 1:
            reveal_type(self.thing)

        local = self.thing
        if local.type == 1:
            reveal_type(local)
```
```sh
test.py:14:25 - information: Type of "self.thing" is "A | B"
test.py:18:25 - information: Type of "local" is "B"
```

**VS Code extension or command-line**
cli, 1.1.272

**Additional context**
Mypy output:
```sh
$ mypy test.py
test.py:14: note: Revealed type is "test.B"
test.py:18: note: Revealed type is "test.B"
```

Strangely enough, pyright correctly narrows both to `B` if there's a condition using `self.thing` anywhere in the method, e.g.
```py
    def test(self) -> None:
        if self.thing.type == 1:
            reveal_type(self.thing)

        if self.thing: pass

        local = self.thing
        if local.type == 1:
            reveal_type(local)
```

**Comment by erictraut** [comment]
Thanks for the bug report. This will be fixed in the next release.

**Comment by erictraut** [comment]
I needed to revert this bug fix because it exposed some other issues that need to be addressed first. The underlying cause of this bug was masking some other issues.

**Comment by erictraut** [comment]
I fixed the other bugs that this fix uncovered and un-reverted the change. This will be addressed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.274, which I just published. It will also be included in a future release of pylance.
